library/ziti.c

```diff
@@ -180,19 +184,61 @@ int NF_init(const char* config, uv_loop_t* loop, nf_init_cb init_cb, void* init_
             time_str, start_time.tv_usec/1000);
 
     PREP(ziti);
+
+    if (opts->config == NULL && (opts->controller == NULL || opts->tls == NULL)) {
+        ZITI_LOG(ERROR, "config or controller/tls has to be set");
+        return ZITI_INVALID_CONFIG;
+    }
+
     nf_config *cfg = NULL;
-    tls_context *tls = NULL;
+    if (opts->config != NULL) {
+        TRY(ziti, load_config(opts->config, &cfg));
+    }
+    if (opts->controller == NULL) {
+        opts->controller = strdup(cfg->controller_url);
```

Comment:
i don't see opts->controller getting free'ed in here? either i missed it or it's done somewhere else so i'm just commenting wondering if that's the case?